### ISSUE TITLE
JSON RPC net_version and eth_chainId sometimes fail when running a localhost fork

### ISSUE THREAD

**Issue by ajb413** [nomiclabs/hardhat#1534]
I've been really struggling to pinpoint the cause of this issue; I'm not 100% certain that it is Hardhat but I am leaning towards it. I am running tests against a localhost fork of mainnet at a specific block. The project is Compound.js, which is an SDK written in TypeScript that wraps around ethers.js.

The issue is that calls to `net_version` seem to work when I call them from within a test, but they are breaking when an instance of Compound.js SDK calls it.

Breaks:
```js
const compound = new Compound('http://localhost:8545', { privateKey });
const trx = await compound.repayBorrow('USDC', 1, '0xbadaddress', false); // a different error should be thrown

// for the `trx` line, an ethers.js error gets thrown:
//     reason: 'missing response',
//     code: 'SERVER_ERROR',
//     requestBody: '{"method":"net_version","id":42,"jsonrpc":"2.0"}',
//     serverError: Error: connect ECONNREFUSED 127.0.0.1:8545
```

The culprit (I think) is [this one line of code](https://github.com/compound-finance/compound-js/blob/master/src/eth.ts#L227), because it is the only place in the SDK where `net_version` is explicitly called.

```ts
networkId = await _provider.send('net_version');
```

I tried switching it to `eth_chainId`, but I get the same error except for `eth_chainId`.

```diff
- networkId = await _provider.send('net_version');
+ networkId = (await _provider.send('eth_chainId')).chainId;
```

I tried calling `net_version` explicitly within the same test, and it works fine (same with `eth_chainId`).

Works fine (both console.logs spit out a 1):
```js
const p = new ethers.providers.JsonRpcProvider('http://localhost:8545');
const ethersNetVersion = await p.send('net_version');
console.log('ethers net_version', +ethersNetVersion);
const hreNetVersion = await hre.network.provider.send('net_version');
console.log('hre net_version', +hreNetVersion);
```

I try doing both of these things within a single test. The error occurs for all of my tests that use an instance of `compound`, not just this one test.

**The reason I do not think that the cause is ethers.js or the Compound.js SDK is because this call works perfectly fine if I use a ganache-core fork of mainnet.**

Theories:
- There is something fishy with the version of ethers that gets used in the hardhat test environment. I tried upgrading ethers.js but I get the same issue. And I suspect that using the non-special version of ethers does not work out with the hardhat localhost instance.
- Some issue with typescript? The call comes from within a ts file.

Steps to reproduce (need **your own** archive node access):
```
## first set an environment variable to your own JSON RPC URL with archive node access
export MAINNET_PROVIDER_URL="https://eth-mainnet.alchemyapi.io/v2/your-api-key"

git clone https://github.com/compound-finance/compound-js.git
cd compound-js
git checkout -b hardhat-tests
git pull origin hardhat-tests
npm i
npm test -- -g 'fails cToken.repayBorrow behalf address'
```

What this does is it [runs this single test](https://github.com/compound-finance/compound-js/blob/hardhat-tests/test/cToken.test.js#L422) that has the combined working and broken code snippets above.

If anyone can pinpoint the issue, identify the bug, or at least point me in the right direction I would really appreciate it!

**Comment by fvictorio** [comment]
Hey @ajb413, I know this is a silly question but... are you sure something is running in `localhost:8545`? That error typically shows up when that port isn't open. And the reason it might work with ganache is that ganache always (as far as I know) starts an HTTP server, while hardhat doesn't (for example, when you run your tests without passing any `--network`, you are using an in-process hardhat network that doesn't open any port).

In other words, that should work if you've previously run `npx hardhat node` in another terminal. Although most of the time you don't need to do that and you just use the injected provider that is connected to the in-process network.

**Comment by ajb413** [comment]
If I comment out my line `hre.run('node')` in my [main test file](https://github.com/compound-finance/compound-js/blob/hardhat-tests/test/index.js#L40) and run `npx hardhat node` in a separate command line, the test succeeds, even with both of the different objects making those same requests.

I am guessing that something is off with my `hre.run('node')` line, which I highly prefer to use instead of a separate `npx hardhat node`. I am having trouble finding docs for it, specifically what parameters I can pass to the command in the second parameter.

Is there a way to run my tests properly without having to run hardhat in a separate command line? The log created by `hre.run('node')` that a server is running at 8545 is at the least deceiving if some objects can query it but others can't. If that is actually the intended behavior, which I have a feeling it isn't, some docs explaining why would be great.

**Comment by fvictorio** [comment]
Oh, I didn't realize you were programmatically starting the node task. I'll check this again later, thanks for clarifying.

**Comment by hoont** [comment]
I just went to report this issue to see it was the most recent. I'm having the exact same problem, with slightly different circumstances - I have some small scripts that muck around with Uniswap contracts on a forked hardhat mainnet node that have all randomly stopped working today. Stdout from the scripts prints

`amountsout error : ProviderError: Invalid signature v value`

and the forked node shows the error occurs in `eth_call`?

![image](https://user-images.githubusercontent.com/12984183/121614766-be3b4480-caa2-11eb-8d17-00d013de53db.png)

This is all I could find

**Comment by fvictorio** [comment]
@hoont that seems like a different issue. If you can make a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example), please open a new issue with a link to it.

**Comment by ajb413** [comment]
@fvictorio Is there a way to configure a hh node from within a script to function exactly the same as doing `npx hardhat node`?

So I would instead do something like this:

`node myLocalHardhat.js`

myLocalHardhat.js:
```js
const hre = require('hardhat');
hre.run('node');
// do some configuring here?
```

Perhaps, if there is a way to do that, it can be my workaround for now. Thank you!

**Comment by alcuadrado** [comment]
Do you mean passing params as you would in the cli? `run` takes a second object that does that.

**Comment by ajb413** [comment]
@alcuadrado I'm interested in doing that if the script-made node behaves exactly like the node that is launched directly from the cli, so my tests will pass as expected. To summarize the above:

Using `hre.run('node');` from within a script: test script fails
Instead using `npx hardhat node` from the cli: test script passes

I was asking, is there a way to pass params to `hre.run('node');` so that it behaves exactly like `npx hardhat node` so my tests will pass?

If `hre.run('node');` and `npx hardhat node` *are* supposed to function identically, then I assume this must be due to a bug? I could be wrong.

**Comment by alcuadrado** [comment]
The cli does `hre.run("node")` internally. I suspect you are not waiting for the RPC server to launch.

**Comment by ajb413** [comment]
> The cli does `hre.run("node")` internally. I suspect you are not waiting for the RPC server to launch and be ready to accept connections. It takes a few millis. Take a look at how the `node` task does it.

How long should I wait? Is 5 seconds too brief? https://github.com/compound-finance/compound-js/blob/hardhat-tests/test/index.js#L40

**Comment by fvictorio** [comment]
@ajb413 I haven't been able to look into this again, but something you can try is using the node subtask, which gives you a server instance. That way you'll have better control over what's going on. It's used like this:

```js
const { TASK_NODE_CREATE_SERVER } = require("hardhat/builtin-tasks/task-names");

const server = await hre.run(TASK_NODE_CREATE_SERVER, {
  hostname: "localhost",
  port: 8545,
  provider: hre.network.provider,
});

await server.listen()

// server is ready here

await server.close()
```

**Comment by ajb413** [comment]
@fvictorio Hi Franco, thank you for continuing to help me out debugging this. I tried this too and I am still seeing the same error.

**Comment by ajb413** [comment]
@alcuadrado @fvictorio I've tried several things, one of which is converting all the tests and config to Typescript. I am suspicious of the injected hre.ethers vs importing normal ethers. I think that perhaps there is something fishy in there? Do you have any suggestions for me to dig in debugging this issue?

All of these tests pass normally if I use ganache core (with the exception of one, which is a completely unrelated ganache bug - which is also the main reason I switched to hardhat).

**Comment by fvictorio** [comment]
Hey @ajb413, sorry, I haven't had time to look into this yet, but I'll try to do it this week.

**Comment by fvictorio** [comment]
Hey @ajb413, sorry again for not being able to look into this before. I think I see what your issue is.

You are creating the `Compound` object here:

https://github.com/compound-finance/compound-js/blob/99c7e5d9a72daa2704b8dfda036e808112930787/test/cToken.test.js#L12

The function exported by that module is called by this `describe`:

https://github.com/compound-finance/compound-js/blob/99c7e5d9a72daa2704b8dfda036e808112930787/test/index.js#L47

And the body of a `describe` is called before the `before` hooks are called. And since the `Compound` constructor does this:

https://github.com/compound-finance/compound-js/blob/99c7e5d9a72daa2704b8dfda036e808112930787/src/index.ts#L76

Then the `net_version` RPC method is called before the node is initialized.

My guess is that moving the creation of the `Compound` instances to a `before` hook should fix this.

I'm going to close this now since it doesn't seem to be a problem with Hardhat itself, but please re-open if you think it is.

**Comment by ajb413** [comment]
Nailed it! Thank you so much for your help! I did not suspect that that mattered. I made that change and now it works.
